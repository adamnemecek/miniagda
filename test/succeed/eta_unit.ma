-- 2009-06-25 eta expansion for the unit type

data Unit : Set
{
  unit : Unit
}

fun P : Unit -> Set
{
  P unit = Unit
}

fun p : (u : Unit) -> P u
{
  p x = unit
}

fun q : (u : Unit) -> P u
{
  q unit = unit
}

-- what also should work is
-- q .unit = unit

-- 2009-09-19

data Bool : Set
{ true  : Bool
; false : Bool
}

let r' : Bool -> Unit
       = \ b -> unit

let pr' : (b : Bool) -> P (r' b)
       = \ b -> unit

fun r : Bool -> Unit
{ r true = unit
; r false = unit
}

-- definitions need also to be eta-expanded
-- otherwise the following does not typecheck
let pr : (b : Bool) -> P (r b)
       = \ b -> unit
