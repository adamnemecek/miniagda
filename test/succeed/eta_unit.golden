--- opening "eta_unit.ma" ---
--- scope checking ---
--- type checking ---
type  Unit : Set
term  Unit.unit : < Unit.unit : Unit >
type  P : Unit -> Set
{ P un!t = Unit
}
term  p : (u : Unit) -> P u
{ p x = Unit.unit
}
term  q : (u : Unit) -> P u
{ q un!t = Unit.unit
}
type  Bool : Set
term  Bool.true : < Bool.true : Bool >
term  Bool.false : < Bool.false : Bool >
term  r' : Bool -> Unit
term  r' = \ b -> Unit.unit
term  pr' : (b : Bool) -> P (r' b)
term  pr' = \ b -> Unit.unit
term  r : Bool -> Unit
{ r Bool.true = Unit.unit
; r Bool.false = Unit.unit
}
term  pr : (b : Bool) -> P (r b)
term  pr = \ b -> Unit.unit
--- evaluating ---
--- closing "eta_unit.ma" ---
